### FUNGSI dalam Python

#### Pengertian dan sintak

Dalam penulisan kode program kita dapat memisahkan bagian yang merupakan garis besar dari bagian yang merupakan detil implementasi. Pemisahan seperti ini menghasilkan struktur yang lebih jelas yaitu dengan dekomposisi, membagi kode menjadi modul-modul, beberapa sub-rutin. Modul-modul tersebut diberi nama dan dipisahkan dari garis besar kode sehingga detilnya tersembunyi atau terabstraksi.
Modul-modul itu dapat dituangkan dalam bentuk fungsi atau kelas. Kali ini pembahasan kita adalah mengenai fungsi.   
Sintak fungsi ialah  
    
**def** _nama_fungsi_ **(parameter,parameter):**    
- *badan fungsi dengan identasi (default: 4 spasi)*    
**return**

Ketika interpreter membaca definisi fungsi, maka disiapkan (setup) variabel parameter-parameter, lalu ketika dipanggil maka argumen dari pemanggil dioperkan ke dalam fungsi.    

Ciri-ciri fungsi dalam Python yaitu:
   - tidak diperlukan deklarasi tipe output fungsi seperti _void, int, string_
   - tidak diperlukan deklarasi tipe parameter/argumen input seperti _int, string_
   - blok/grup badan fungsi ditandai dengan indentasi dan diawali titik dua pada akhir kalimat definisi, *bukan dengan { }*
   - doc string dokumentasi merupakan bagian dari sintaks fungsi, yang dapat ditampilkan juga dengan help()

In [1]:
def f():
    pass

print(type(f))

<class 'function'>


In [2]:
def nama_sesuai_aturan(nama):
    """ Memeriksa apakah argumen nama telah memenuhi beberapa aturan yang baru. 
    input berupa string
    output berupa tuple (boolean dan list keterangan)
    """
    cek = True
    keterangan =[]
    if len(nama)> 60:
        cek = False
        ketereangan.append("terlalu panjang")
    if len(nama)<5:
        cek = False
        keterangan.append("terlalu singkat")
    if len(nama.split(" "))<2:
        cek = False
        keterangan.append("minimal 2 kata")
    if any(c.isdigit() for c in nama):
        cek = False
        keterangan.append("dilarang menggunakan angka")
    return cek, keterangan

def main():
    """ Program utama """
    nama_nama = "Aa I", "Anakku_sayang", "Raden-3", "Joko Wiseso Suseno Pradjaningrat Kusumabangsa"
    for nama in nama_nama:
        cek, ket = nama_sesuai_aturan(nama)
        if cek:
            print(f"'{nama}' sudah sesuai aturan")
        else:
            print(f"'{nama}' tidak sesuai karena {ket}")

if __name__ == '__main__':
    main()

'Aa I' tidak sesuai karena ['terlalu singkat']
'Anakku_sayang' tidak sesuai karena ['minimal 2 kata']
'Raden-3' tidak sesuai karena ['minimal 2 kata', 'dilarang menggunakan angka']
'Joko Wiseso Suseno Pradjaningrat Kusumabangsa' sudah sesuai aturan


Modul fungsi biasanya terdiri atas beberapa kalimat kode, sehingga memberikan efek abstraksi jika digantikan dengan kalimat panggilan fungsi.   
Di dalam definisi fungsi dapat didefinisikan atau memanggil fungsi lain yang lebih spesifik/detil.
Urutan penulisan kode program harus 'terbalik' yakni kalimat definisi fungsi harus mendahului kalimat pemanggilannya karena interpreter Python yang membaca baris per baris dari atas akan segera menampilkan tampilan kesalahan jika tidak/belum menemukan definisi fungsi yang dipanggil.  


In [3]:
help(nama_sesuai_aturan)
print(f'{type(nama_sesuai_aturan)= }')
print(f'{callable(nama_sesuai_aturan)= }')

Help on function nama_sesuai_aturan in module __main__:

nama_sesuai_aturan(nama)
    Memeriksa apakah argumen nama telah memenuhi beberapa aturan yang baru. 
    input berupa string
    output berupa tuple (boolean dan list keterangan)

type(nama_sesuai_aturan)= <class 'function'>
callable(nama_sesuai_aturan)= True


#### Fungsi Bawaan (built-in)    
Pustaka Python bawaan yang terintegrasi memiliki fungsi fungsi sebagai berikut

In [4]:
# memilih fungsi bawaan dari dir(__builtins__)
bf = [fx for fx in dir(__builtins__) if '_' not in fx and 'builtin_function' in str(type(eval(fx)))]
print(bf)

['abs', 'aiter', 'all', 'anext', 'any', 'ascii', 'bin', 'breakpoint', 'callable', 'chr', 'compile', 'delattr', 'dir', 'divmod', 'eval', 'exec', 'format', 'getattr', 'globals', 'hasattr', 'hash', 'hex', 'id', 'isinstance', 'issubclass', 'iter', 'len', 'locals', 'max', 'min', 'next', 'oct', 'open', 'ord', 'pow', 'print', 'repr', 'round', 'setattr', 'sorted', 'sum', 'vars']


#### Argumen

Data informasi dapat dioper ke dalam fungsi berupa argumen kepada variabel parameter yang tertera dalam kurung setelah nama identifier fungsi, seperti variabel nama pada contoh di atas.    
Jumlah argumen yang akan dioper harus sama dengan jumlah parameter yang tertera, bahkan urutannya pun harus sesuai.    
Tetapi jika diperlukan maka aturan jumlah dapat diganti dengan aturan arbitrary arguments atau argumen sembarang jumlahnya dengan menggunakan simbol satu asterisk \*args. Demikian juga aturan urutan dapat digantikan dengan aturan penggunaan keyword arguments sebanyak diperlukan atau sembarang jumlahnya menggunakan dua asterisk \*\*kwargs. 

In [5]:
# jumlah argument harus cocok dengan parameter penampungnya
def daftar(nomor, barang):
    print(f'No.{nomor} {barang.title()}')

punyaku=('laptop','lampu meja', 'hape')
for x,y in enumerate(punyaku,1):
    daftar(x,y)

No.1 Laptop
No.2 Lampu Meja
No.3 Hape


In [6]:
# argumen sembarang jumlahnya menggunakan *args
def nama2cu2(*args_nama):
    for name in args_nama:
        print(name)
    print(f'yang bungsu ialah {args_nama[-1]}')
    
nama2cu2('Januar','Maulana','Jasmine')
    

Januar
Maulana
Jasmine
yang bungsu ialah Jasmine


In [7]:
# argumen pakai nama, urutan bebas, argumen opsional dengan default, **kwargs
def bunga(nama, warna='menarik', **lainnya):
    match nama:
        case 'anggrek':
            print(f'bunga anggrek warna {warna}')
        case 'gelombang cinta':
            print(f'berdaun hijau bergelombang')
        case _:
            print(f'{nama} berwarna {warna} {lainnya}')
            
bunga('anggrek','putih')
bunga(warna='ungu', nama='anggrek')
bunga('gelombang cinta')
bunga(warna='pink', nama='bunga bulan desember', harga='200K', stock=5)

bunga anggrek warna putih
bunga anggrek warna ungu
berdaun hijau bergelombang
bunga bulan desember berwarna pink {'harga': '200K', 'stock': 5}


#### Parameter opsional serta nilai default - kasus rekursi deret Fibonacci

Pada contoh di atas parameter warna dapat menerima argumen atau menggunakan nilai default seperti tertera jika tidak mendapat operan argumen ketika fungsi dipanggil. Artinya dalam kasus default tidak perlu menyertakan argumen.     
Saat interpreter membaca statement definisi fungsi, maka parameter yang tertera disetup agar siap menerima argumen. Untuk parameter opsional disiapkan prosedur tambahan yaitu jika tidak menerima argumen (berdasarkan posisi urutan atau nama parameter), maka nilainya diambilkan dari suatu lokasi yang ditentukan yang berisi nilai default yang tertera pada kalimat definisi. Ketika fungsi dipanggil, proses setup atau inisialisasi ini tidak diulang lagi.    
Kita bahas fitur ini karena biasa digunakan sebagai memo untuk mengingat sesuatu hasil dalam fungsi rekursi agar kinerjanya tinggi.

In [8]:
# default parameter and local variable
def f(memo='awal'):
    if 'awal' == memo:
        memo = 'berikutnya'
    print(memo)

print(f.__defaults__)   # parameter fungsi sudah selesai disetup, ini lokasi untuk nilai default
f()
f()
f()

('awal',)
berikutnya
berikutnya
berikutnya


In [9]:
# default parameter and mutable parameter
def f(memo=[]):
    memo.append("*")    # anggota list memo ditambah
    print(f'{memo = }')

print(f.__defaults__)   # parameter fungsi sudah selesai disetup, ini lokasi untuk nilai default
f()
f()
f()

([],)
memo = ['*']
memo = ['*', '*']
memo = ['*', '*', '*']


In [10]:
# memo tersimpan bersama scope f karena merupakan atribut dari fungsi f
print(f.__defaults__)
f()
f.__defaults__[0].clear()
print(f.__defaults__)
f()

(['*', '*', '*'],)
memo = ['*', '*', '*', '*']
([],)
memo = ['*']


Dalam contoh fungsi rekursi menggunakan memo akan terlihat manfaat penggunaan cara ini.def fib(n):
if n in (0, 1):
return n
return fib(n - 2) + fib(n - 1)

In [11]:
# rekursi naif untuk deret Fibonacci
def fib(n):
    if n in (0, 1):
        return n
    return fib(n - 1) + fib(n - 2)
print(fib(12))

144


In [12]:
def fib_debug(n: int) -> int:
    if n in (0, 1):
        print('.', end=" ")
        return n
    print(n, end=' ')
    return fib_debug(n - 1) + fib_debug(n - 2)
fib_debug(12)

12 11 10 9 8 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 8 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 9 8 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 10 9 8 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 8 7 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 5 4 3 2 . . . 2 . . 3 2 . . . 6 5 4 3 2 . . . 2 . . 3 2 . . . 4 3 2 . . . 2 . . 

144

Contoh di atas memperlihatkan parahnya evaluasi berulang-ulang untuk argumen yang sama. Jika dihitung maka untuk menghasilkan hasil fib(12) digunakan iterasi sebanyak fib(13)-1 kali, bodoh sekali. Makanya pada argumen yang lebih besar, fib(34) yang menghasilkan angka 5.702.887 akan dilakukan iterasi evaluasi sebanyak 9.227.464 kali. Untuk angka lebih besar dari 34, komputer Anda akan mulai terasa macet.

In [13]:
# cached fibonacci "store results in default argument"
def fast_fib(n, memo = {0: 0, 1: 1}):
    if n in memo:
        return memo[n]  # get element values from default argument object
    print(n, end=',') 
    memo[n] = fast_fib(n-1) + fast_fib(n-2)  # iterate & save to container’s element
    return memo[n]


print('fibonacci tanpa kalkulasi ulang sama sekali')
numbers = 30,40, 41, 42
for num in numbers:
    print(f' -> fibonacci({num}) = {fast_fib(num):,}')


# fast_fib.__defaults__ = ({0:0, 1:1},) # inisialisasi ulang memo jika diperlukan

fibonacci tanpa kalkulasi ulang sama sekali
30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2, -> fibonacci(30) = 832,040
40,39,38,37,36,35,34,33,32,31, -> fibonacci(40) = 102,334,155
41, -> fibonacci(41) = 165,580,141
42, -> fibonacci(42) = 267,914,296


Jika inisialisasi ulang memo ditutup, seperti pada cell di atas, maka hasil run seperti pada cell berikut, tanpa run definisi fungsi lagi, akan menghasilkan nilai output tanpa perhitungan sama sekali, karena semua hasilnya diambilkan dari memo

In [14]:
print('fibonacci tanpa menghitung sama sekali, sudah ada dalam memo')
numbers = 30,40, 41, 42
for num in numbers:
    print(f'fibonacci({num}) = {fast_fib(num):,}')
print("\nmemo:\n",fast_fib.__defaults__[0])

fibonacci tanpa menghitung sama sekali, sudah ada dalam memo
fibonacci(30) = 832,040
fibonacci(40) = 102,334,155
fibonacci(41) = 165,580,141
fibonacci(42) = 267,914,296

memo:
 {0: 0, 1: 1, 2: 1, 3: 2, 4: 3, 5: 5, 6: 8, 7: 13, 8: 21, 9: 34, 10: 55, 11: 89, 12: 144, 13: 233, 14: 377, 15: 610, 16: 987, 17: 1597, 18: 2584, 19: 4181, 20: 6765, 21: 10946, 22: 17711, 23: 28657, 24: 46368, 25: 75025, 26: 121393, 27: 196418, 28: 317811, 29: 514229, 30: 832040, 31: 1346269, 32: 2178309, 33: 3524578, 34: 5702887, 35: 9227465, 36: 14930352, 37: 24157817, 38: 39088169, 39: 63245986, 40: 102334155, 41: 165580141, 42: 267914296}


#### Pustaka functools

Setelah memahami persoalan efisiensi dan solusi memoize pada fungsi rekursi, maka kita dapat mengapresiasi pustaka *perkakas-fungsi* **functools** dengan dekorator cache yang memungkinkan tercapainya abstraksi dan keringkasan maksimal seperti pada contoh berikut.

In [15]:
# rekursi menggunakan dekorator cache sebagai memoize
from functools import cache

@cache
def fibonacci(n):
    return fibonacci(n-1) + fibonacci(n-2) if n>1 else n

print(f'{fibonacci(35) = }')


fibonacci(35) = 9227465


Contoh kode program di atas dapat dijelaskan dengan closure dan decorator sebagai berikut:

In [16]:
# memoize 
def memoize(f):
    memo = {}        # ini yang disebut free-variable

    def closure(n):  # ini yang disebut closure
        if n not in memo:
            memo[n] = f(n)
        return memo[n]

    return closure

@memoize  # decorator: syntatic sugar for fib = memoize(fib)
def fib(n):
    return fib(n - 1) + fib(n - 2) if n>1 else n
## fib = memoize(fib)  # decorator dalam bentuk aslinya diletakkan disini setelah definisi

print(f'{fib(35) = }') 

print(fib.__code__.co_freevars[1])
print(fib.__closure__[1].cell_contents)
print('=-'*50,'\n')

fib(35) = 9227465
memo
{1: 1, 0: 0, 2: 1, 3: 2, 4: 3, 5: 5, 6: 8, 7: 13, 8: 21, 9: 34, 10: 55, 11: 89, 12: 144, 13: 233, 14: 377, 15: 610, 16: 987, 17: 1597, 18: 2584, 19: 4181, 20: 6765, 21: 10946, 22: 17711, 23: 28657, 24: 46368, 25: 75025, 26: 121393, 27: 196418, 28: 317811, 29: 514229, 30: 832040, 31: 1346269, 32: 2178309, 33: 3524578, 34: 5702887, 35: 9227465}
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=- 



#### Fungsi Rekursi Ekor (Tail Recursion)

Rekursi ekor ialah rekursi yang menyimpan perhitungan hasil sementara dalam parameternya, sehingga saat mencapai titik balik (base case) langsung dapat mengeluarkan hasilnya tidak perlu mundur kembali lagi untuk menghitung formula yang tertunda. Perhatikan bedanya pada cell berikut.

In [17]:
# rekursi kepala
def factorial(n):
    return 1 if n==1 else n* factorial(n-1)

# rekursi ekor
def factorial(n, ans=1):
    return ans if n==1 else factorial(n-1, n*ans)

print(f'{factorial(5)=}') 

factorial(5)=120


Gambaran proses eksekusi rekursi kepala dan rekursi ekor.
<pre>Rekursi kepala factorial(5) sebagai berikut:
  5 * factorial(4)
      4 * factorial(3)
          3 * factorial(2)
              2 * factorial(1)
                  1
                  mundur ke stack sebelumnya
              2 * 1 = 2
          3 * 2 = 6
      4 * 6 = 24
  5 * 24 = 120
  return 120
   
Rekursi ekor factorial(5) sebagai berikut:<pre>
  factorial(5, 1) = 
     factorial(4, 5*1=5) =
        factorial(3, 4*5=20) =
           factorial(2, 3*20=60) =
              factorial(1, 2*60=120) =
              120
  return 120
</pre>


#### Perencanaan Dinamis (Dynamic programming)